packages/notifier/src/storesub.js

```diff
@@ -2,8 +2,85 @@
 import { E } from '@endo/eventual-send';
 import { Far, makeMarshal } from '@endo/marshal';
 import { observeIteration } from './asyncIterableAdaptor.js';
+import { makePublishKit, subscribeEach } from './publish-kit.js';
 import { makeSubscriptionKit } from './subscriber.js';
 
+/**
+ * Begin iterating the source, storing serialized iteration values.  If the
+ * storageNode's `setValue` operation rejects, the iteration will be terminated.
+ *
+ * Returns a StoredSubscription that can be used by a client to directly follow
+ * the iteration themselves, or obtain information to subscribe to the stored
+ * data out-of-band.
+ *
+ * @template T
+ * @param {Subscriber<T>} subscriber
+ * @param {ERef<StorageNode>} [storageNode]
+ * @param {ERef<ReturnType<typeof makeMarshal>>} [marshaller]
+ * @returns {StoredSubscriber<T>}
+ */
+export const makeStoredSubscriber = (
+  subscriber,
+  storageNode,
+  marshaller = makeMarshal(undefined, undefined, {
+    marshalSaveError: () => {},
+  }),
+) => {
+  /** @type {Unserializer} */
+  const unserializer = Far('unserializer', {
+    unserialize: E(marshaller).unserialize,
+  });
+
+  // Abort the iteration on the next observation if the publisher ever fails.
+  let publishFailed = false;
+  let publishException;
+
+  const fail = err => {
+    publishFailed = true;
+    publishException = err;
+  };
+
+  // Must *not* be an async function, because it sometimes must throw to abort
+  // the iteration.
+  const publishValue = obj => {
+    assert(storageNode);
+    if (publishFailed) {
+      // To properly abort the iteration, this must be a synchronous exception.
+      throw publishException;
+    }
+
+    // Publish the value, capturing any error.
+    E(marshaller)
+      .serialize(obj)
+      .then(serialized => {
+        const encoded = JSON.stringify(serialized);
+        return E(storageNode).setValue(encoded);
+      })
+      .catch(fail);
+  };
+
+  if (storageNode) {
+    // Start publishing the source.
+    observeIteration(subscribeEach(subscriber), {
+      updateState: publishValue,
+      finish: publishValue,
+    }).catch(fail);
+  }
+
+  /** @type {StoredSubscriber<T>} */
+  const storesub = Far('StoredSubscriber', {
+    ...subscriber,
+    getStoreKey: () => {
+      if (!storageNode) {
+        return {};
+      }
+      return E(storageNode).getStoreKey();
```

Comment:
do you mean you'd expect this to return falsy when storageNode is falsy? agreed. I tried returning null but ran into a type error.

I'll leave this conversation open to come back to it. Let me know if you're talking about somethign else

Reply:
That, and also that a falsy storageNode is accepted while a storageNode that resolves to a falsy value (e.g., `Promise.resolve(undefined)`) results in an exception. I think this also intersects my concern about allowing invocation without a valid storage node, which seems to unnecessarily complicate things (e.g., note how the output from `getStoreKey` is meaningless in such a configuration).

Reply:
I agree with the perspective. In production, the storageNode and marshaller should always be there so it would be fine to require it. In tests, they aren't always. I think it would be better to solve that, but pragmatically I'm not sure it's worth it.

If I can't remove the kludge in the tests, I'll at least remove it from `@agoric/notifier` and make the contracts deal with it.

Reply:
> In production, the storageNode and marshaller should always be there

We have the sim chain, which while perhaps not production, is supported for smart contract development work. In the sim chain, there's no chainStorage.

Reply:
Thanks, that's the thing I was trying to remember. Since it's a "sim" chain, would it be reasonable to _simulate_ chainStorage? E.g. in-memory stoageNode, identity marshaller.

Reply:
Made it required and updated bootstrap to always provide it [cbaa70b](https://github.com/Agoric/agoric-sdk/pull/5704/commits/cbaa70b6a3f7498c18f26aa85e8ee4a80088d797), falling back to a no-op "NullStorageNode" when ChainStorage isn't available.